File: android/app/src/main/java/com/sensorysymphony/storymode/Extract.kt
package com.sensorysymphony.storymode

import org.json.JSONArray

/** Mirrors engine/extract.js — chunk answers, pull grounded factoids. Ungrounded output is dropped. */
object Extract {

    fun chunk(text: String, target: Int = 120): List<String> {
        val paras = text.split(Regex("\n+")).map { it.trim() }.filter { it.isNotEmpty() }
        val chunks = mutableListOf<String>()
        var cur = mutableListOf<String>()
        var count = 0
        fun flush() { if (cur.isNotEmpty()) { chunks.add(cur.joinToString("\n")); cur = mutableListOf(); count = 0 } }
        for (p in paras) {
            val w = p.split(Regex("\\s+")).size
            if (w > target * 1.8) {
                flush()
                var scur = mutableListOf<String>(); var scount = 0
                val sentences = Regex("[^.!?]+[.!?]+|\\S[^.!?]*$").findAll(p).map { it.value }.toList().ifEmpty { listOf(p) }
                for (s in sentences) {
                    val sw = s.split(Regex("\\s+")).size
                    if (scount + sw > target && scur.isNotEmpty()) { chunks.add(scur.joinToString(" ")); scur = mutableListOf(); scount = 0 }
                    scur.add(s.trim()); scount += sw
                }
                if (scur.isNotEmpty()) chunks.add(scur.joinToString(" "))
                continue
            }
            if (count + w > target && cur.isNotEmpty()) flush()
            cur.add(p); count += w
        }
        flush()
        return chunks.ifEmpty { listOf(text) }
    }

    /** Anti-invention armor: the verbatim must actually appear in the source chunk. */
    fun grounded(verbatim: String, source: String): Boolean {
        val v = StoryState.norm(verbatim); val s = StoryState.norm(source)
        if (v.isEmpty()) return false
        if (s.contains(v)) return true
        val vt = v.split(" ").filter { it.length > 2 }
        if (vt.isEmpty()) return false
        val st = s.split(" ").toSet()
        return vt.count { it in st }.toDouble() / vt.size >= 0.8
    }

    data class Extracted(val category: String, val text: String, val verbatim: String, val weight: Double, val flags: List<String>)

    suspend fun fromAnswer(llm: LlmBridge, systemPrompt: String, question: String, answer: String,
                           onProgress: (Int, Int) -> Unit = { _, _ -> }): List<Extracted> {
        val chunks = chunk(answer)
        val out = mutableListOf<Extracted>()
        chunks.forEachIndexed { i, ch ->
            onProgress(i + 1, chunks.size)
            val user = "INTERVIEW QUESTION THAT PROMPTED THIS:\n$question\n\nCUSTOMER'S WORDS:\n$ch"
            val parsed = llm.chatJson(listOf(Msg("system", systemPrompt), Msg("user", user)), temperature = 0.2) ?: return@forEachIndexed
            val fa = parsed.optJSONArray("factoids") ?: JSONArray()
            for (j in 0 until fa.length()) {
                val f = fa.optJSONObject(j) ?: continue
                val text = f.optString("text"); if (text.isBlank()) continue
                val verbatim = f.optString("verbatim").ifBlank { text }
                if (!grounded(verbatim, ch)) continue
                val flags = mutableListOf<String>()
                val fl = f.optJSONArray("flags") ?: JSONArray()
                for (k in 0 until fl.length()) fl.optString(k)?.let { if (it.isNotBlank()) flags.add(it) }
                out.add(Extracted(f.optString("category"), text, verbatim, f.optDouble("heat", 3.0), flags))
            }
            val name = parsed.optString("name")
            if (name.isNotBlank() && name != "null") {
                out.add(Extracted("identity", "name: $name", name, 8.0, listOf("name")))
            }
        }
        return out
    }
}

File: engine/extract.js
// EXTRACT — chunk an answer, pull grounded factoids with the small model.
// The model quotes and files; it never narrates. Ungrounded output is dropped in code.
"use strict";

const fs = require("fs");
const path = require("path");
const { chatJson } = require("./llm");
const { norm } = require("./state");

const PROMPT = fs.readFileSync(path.join(__dirname, "..", "prompts", "extract.txt"), "utf8");

// Paragraph-aware chunking, ~target words per chunk. Small inputs = small-model food.
function chunk(text, target = 120) {
  const paras = String(text).split(/\n{1,}/).map((p) => p.trim()).filter(Boolean);
  const chunks = [];
  let cur = [];
  let count = 0;
  const flush = () => { if (cur.length) { chunks.push(cur.join("\n")); cur = []; count = 0; } };
  for (const p of paras) {
    const w = p.split(/\s+/).length;
    if (w > target * 1.8) {
      // Giant paragraph: split on sentences.
      flush();
      let scur = [], scount = 0;
      for (const s of p.match(/[^.!?]+[.!?]+|\S[^.!?]*$/g) || [p]) {
        const sw = s.split(/\s+/).length;
        if (scount + sw > target && scur.length) { chunks.push(scur.join(" ")); scur = []; scount = 0; }
        scur.push(s.trim()); scount += sw;
      }
      if (scur.length) chunks.push(scur.join(" "));
      continue;
    }
    if (count + w > target && cur.length) flush();
    cur.push(p); count += w;
  }
  flush();
  return chunks.length ? chunks : [String(text)];
}

// Grounding check: the verbatim must actually appear in the source chunk.
// Anti-invention armor — if the model made it up, it does not enter the construct.
function grounded(verbatim, source) {
  const v = norm(verbatim), s = norm(source);
  if (!v) return false;
  if (s.includes(v)) return true;
  // Fuzzy: 80% of the verbatim's tokens appear in order-agnostic form in the source.
  const vt = v.split(" ").filter((w) => w.length > 2);
  if (!vt.length) return false;
  const st = new Set(s.split(" "));
  const hits = vt.filter((w) => st.has(w)).length;
  return hits / vt.length >= 0.8;
}

async function extractFromAnswer(model, question, answer, onProgress) {
  const chunks = chunk(answer);
  const out = [];
  for (let i = 0; i < chunks.length; i++) {
    if (onProgress) onProgress(i + 1, chunks.length);
    const user = `INTERVIEW QUESTION THAT PROMPTED THIS:\n${question}\n\nCUSTOMER'S WORDS:\n${chunks[i]}`;
    const parsed = await chatJson(model, [
      { role: "system", content: PROMPT },
      { role: "user", content: user },
    ], { temperature: 0.2 });
    if (!parsed || !Array.isArray(parsed.factoids)) continue;
    for (const f of parsed.factoids) {
      if (!f || typeof f.text !== "string") continue;
      const verbatim = typeof f.verbatim === "string" ? f.verbatim : f.text;
      if (!grounded(verbatim, chunks[i])) continue; // dropped: not in the customer's words
      out.push({
        category: f.category,
        text: f.text,
        verbatim,
        weight: Number(f.heat) || 3,
        flags: Array.isArray(f.flags) ? f.flags.filter((x) => typeof x === "string") : [],
      });
    }
    if (parsed.name && typeof parsed.name === "string") {
      out.push({ category: "identity", text: `name: ${parsed.name}`, verbatim: parsed.name, weight: 8, flags: ["name"] });
    }
  }
  return out;
}

module.exports = { chunk, grounded, extractFromAnswer };
